cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -375,10 +379,19 @@ private void publishStopProgram(String stopPayload) {
     ProgramRunId programRunId = new ProgramRunId(programRunInfo.getNamespace(), programRunInfo.getApplication(),
                                                  ProgramType.valueOf(programRunInfo.getProgramType()),
                                                  programRunInfo.getProgram(), programRunInfo.getRun());
-    programStateWriter.killed(programRunId);
+    programStateWriter.stop(programRunId, 0);
     LOG.debug("Published program stop message for run {}", programRunInfo.getRun());
   }
 
+  private void publishKillProgram(String stopPayload) {
+    ProgramRunInfo programRunInfo = GSON.fromJson(stopPayload, ProgramRunInfo.class);
+    ProgramRunId programRunId = new ProgramRunId(programRunInfo.getNamespace(), programRunInfo.getApplication(),
+                                                 ProgramType.valueOf(programRunInfo.getProgramType()),
+                                                 programRunInfo.getProgram(), programRunInfo.getRun());
+    programStateWriter.killed(programRunId);
```

Comment:
This will change the program state without actually killing the program. There a couple ways we could handle kills, let's discuss more and handle it in a separate PR.

Reply:
changed to stop with timeout=0